lib/system.nim

```diff
@@ -4441,6 +4442,14 @@ when defined(genode):
         # and return to thread entrypoint.
 
 
+when defined(nimV2):
+  proc reset*[T](obj: var T) =
+    ## Resets an object `obj` to its initial (binary zero) value.
+    ## Destructor is invoked if T type has destructor.
+    mixin `=destroy`
+    `=destroy`(obj)
+    zeroMem(obj.addr, sizeof(obj))
```

Comment:
Can we add tests on non-seq generic objects in a generic proc or template and also normal object in a block expression.

Requiring mixin here might require every proc that would use `reset` to also have a mixin `=destroy`.
See:
  * https://github.com/nim-lang/Nim/issues/5707 (list comprehensions requires mixin with generics)
  * https://github.com/nim-lang/Nim/issues/6387 Custom seq requires mixin (but iirc seq doesn't)
  * https://github.com/nim-lang/Nim/issues/7385 Mixin in a block expression only works for generic object.

Meta-issue: https://github.com/nim-lang/Nim/issues/8677

Reply:
> Requiring mixin here might require every proc that would use reset to also have a mixin =destroy.

That's not how ``mixin`` works.

Reply:
In generics case it sometimes just does not work, here is another example i'm struggling with: https://github.com/status-im/nim-serialization/issues/4

And the problematic type is a mixin in the template: https://github.com/status-im/nim-serialization/blob/6804ea25372de5919ef87cf13c7544770aacba2f/serialization.nim#L73

Reply:
I don't think mixin comments are relevant . `=destroy` invocation for generic type needs a mixin. Period. There is nothing `reset` specific here.
I have extend the test cover a type with no destructor.

Reply:
What I'm saying is that if we have symbol visibility rules that requires mixin for this proc, we may have symbol visibility issues where mixin doesn't work and the package author will have to mixin `=destroy` in his own proc.